chrome/content/zotero/modules/pluginAPIBase.mjs

```diff
@@ -0,0 +1,383 @@
+/*
+	***** BEGIN LICENSE BLOCK *****
+
+	Copyright © 2024 Corporation for Digital Scholarship
+					 Vienna, Virginia, USA
+					 https://digitalscholar.org
+
+	This file is part of Zotero.
+
+	Zotero is free software: you can redistribute it and/or modify
+	it under the terms of the GNU Affero General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	Zotero is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU Affero General Public License for more details.
+
+	You should have received a copy of the GNU Affero General Public License
+	along with Zotero.  If not, see <http://www.gnu.org/licenses/>.
+
+	***** END LICENSE BLOCK *****
+*/
+
+
+/**
+ * @typedef {object} PluginAPIConfig
+ * @property {string} APIName - The name of the API
+ * @property {string} mainKeyName - The main key in the option object that uniquely identifies it
+ * @property {string} pluginIDKeyName - The key in the option object that identifies the plugin
+ * @property {string} notifyType - The type of the notification to send after an update
+ * @property {string} notifyAction - The action of the notification to send after an update
+ * @property {Record<string, PluginAPIOptDefValue} optionTypeDefinition
+ * - The option validation definition.
+ *
+ *
+ * @typedef {string} PluginAPIOptDefType
+ * - The type of the value
+ *
+ *
+ * @typedef {(value: any) => boolean} PluginAPIOptDefCheckHook
+ * - A function to check the value, returning true if it is valid
+ *
+ *
+ * @typedef {object} PluginAPIOptDefConfig
+ * @property {PluginAPIOptDefType} [type] - The type of the value
+ * @property {boolean} [optional] - Whether the value is optional, default false
+ * @property {PluginAPIOptDefCheckHook} [checkHook]
+ *  - A function to check the value, returning true if it is valid
+ * @property {PluginAPIOptDefConfig} [children]
+ *  - The type definition of the children of the object. The children can be
+ * an object or an array of objects with the same structure
+ * @property {any} [fallbackReturn] - The value to return if the check hook fails
```

Comment:
Rebased, and updated the code to allow APIs to easily configure a fallback return value for functions as hooks when they throw errors.

The API will wrap the function, catch any possible unhandled errors, log them, and use the `fallbackReturn`. It does not promise the value to be correct, it just promises the call does not throw an error.